cloudinit/stages.py

```diff
@@ -692,34 +702,81 @@ def _find_networking_config(self):
+        default_events = get_update_events_config(
+            self.datasource.default_update_events
+        )
+        config_events = self.cfg.get('updates', {})
+        allowed = get_enabled_events(
+            config_events, default_events
+        )
+
+        if not scope:
+            scopes = [allowed.keys()]
+        else:
+            scopes = [scope]
+        LOG.debug('Possible scopes for this event: %s', scopes)
+
+        for evt_scope in scopes:
+            if event_source_type in allowed.get(evt_scope, []):
+                LOG.debug('Event Allowed: scope=%s EventType=%s',
+                          evt_scope, event_source_type)
+                return True
+
+        LOG.debug('Event Denied: scopes=%s EventType=%s',
+                  scopes, event_source_type)
+        return False
+
     def _apply_netcfg_names(self, netcfg):
         try:
             LOG.debug("applying net config names for %s", netcfg)
             self.distro.apply_network_config_names(netcfg)
         except Exception as e:
             LOG.warning("Failed to rename devices: %s", e)
 
+    def _is_first_boot(self):
+        bpath = os.path.join(
+            self.paths.get_cpath(),
+            'sem',
+            'config_scripts_per_once.once'
+        )
+        first_boot = not os.path.exists(bpath)
+        LOG.debug('Is first boot? %s', first_boot)
+        return first_boot
+
     def apply_network_config(self, bring_up):
-        # get a network config
+        """Apply the network config.
+
+        Find the config, determine whether to apply it, apply it via
+        the distro, and optionally bring it up
+        """
+
         netcfg, src = self._find_networking_config()
         if netcfg is None:
             LOG.info("network config is disabled by %s", src)
             return
 
-        # request an update if needed/available
-        if self.datasource is not NULL_DATA_SOURCE:
-            if not self.is_new_instance():
-                if not self.datasource.update_metadata([EventType.BOOT]):
-                    LOG.debug(
-                        "No network config applied. Neither a new instance"
-                        " nor datasource network update on '%s' event",
-                        EventType.BOOT)
-                    # nothing new, but ensure proper names
-                    self._apply_netcfg_names(netcfg)
-                    return
-                else:
-                    # refresh netcfg after update
-                    netcfg, src = self._find_networking_config()
+        current_boot_type = EventType.BOOT
+        if self.is_new_instance():
+            current_boot_type = EventType.BOOT_NEW_INSTANCE
+
+        if not any([
+            self.datasource is NULL_DATA_SOURCE,
+            self.is_new_instance(),  # or should it be self._is_first_boot(),
+            (self.update_event_enabled(current_boot_type, scope='network')
+             and self.datasource.update_metadata([current_boot_type]))
+        ]):
```

Comment:
This section has been where I've been banging my head most against the wall. (With a few exceptions), `apply_networking_config` gets called twice during boot. It get called first in LOCAL mode, and then again in NETWORK mode. 

Previous to this PR, on first boot,  if the datasource got initialized in LOCAL mode, we'd return at this point and not apply networking config in NETWORK mode (`is_new_instance()` returns `False` in network mode on first boot if there's a FILESYSTEM dep). On every subsequent boot, If the BOOT update event is enabled we apply networking config twice (one for LOCAL, one for NETWORK).

That same general behavior should still be true, but applying networking config twice on every boot seems wrong to me. Since `self.is_new_instance()` will always return False on every subsequent boot, that can't be our determining factor. I had considered checking datasource dependencies, and apply during LOCAL if there's a filesystem dependency, and at NETWORK if there's a network dependency, but that could cause problems on first boot because even if we find no local datasource, we still generate the fallback config and use that

Reply:
@TheRealFalcon  just caught up on mail so sorry for the delay.

I think we've had a bug for the longest time on apply_networking_config twice
as you've noticed.  Generally this isn't an issue (other than wasteful) in
that for Datasources that are local and have network-config (say OpenStack)
we already get the network config we're going to use.  In the case that the DS
is network-only (GCE) local never wrote a network config.

That said, I have wanted to apply it only once and the rub has been we've not
decided on how we can keep state between local and init that we've already
rendered network config.  If we want to do so we could use either explicit
state on the cloud object such that when we re-use the object at net time (we
load the object with existing=trust (stages.py:_restore_checked_cache() we'll
have something that can be used to determine if we've already applied.

If we want to avoid attributes on the instance object, we can use ephemeral
files in /run/cloud-init;  this won't have any state implications from boot to
boot but we should carefully think about the case of something re-running
cloud-init from the command line.